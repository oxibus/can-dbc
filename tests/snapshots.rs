#![cfg(all(feature = "serde", feature = "encodings"))]

use nom::Err;
use std::borrow::Cow;
use std::fs;
use std::path::{Path, PathBuf};

use can_dbc::{decode_cp1252, Dbc, Error};
use insta::{assert_debug_snapshot, assert_yaml_snapshot, with_settings};
use test_each_file::test_each_path;

struct TestConfig {
    test_root: &'static str,
    snapshot_suffix: &'static str,
    use_cp1251: bool,
    create_snapshot: bool,
}

static TEST_DIRS: &[TestConfig] = &[
    TestConfig {
        test_root: "shared-test-files",
        snapshot_suffix: "",
        use_cp1251: true,
        create_snapshot: true,
    },
    TestConfig {
        test_root: "opendbc/opendbc/dbc",
        snapshot_suffix: "opendbc",
        use_cp1251: false,
        create_snapshot: false,
    },
];

test_each_path! { for ["dbc"] in "./tests/fixtures/opendbc/opendbc/dbc" as dbc => parse_one_file }
test_each_path! { for ["dbc"] in "./tests/fixtures/shared-test-files" as shared => parse_one_file }
// upper case extension
test_each_path! { for ["DBC"] in "./tests/fixtures/shared-test-files" as shared2 => parse_one_file }

/// Get `test root`, `snapshot name suffix`, `use cp1251`, `create snapshot` for the given path
fn get_test_info(path: &Path) -> Option<(PathBuf, &'static TestConfig)> {
    if !path
        .extension()
        .unwrap_or_default()
        .eq_ignore_ascii_case("dbc")
    {
        return None;
    }
    let path_str = path.to_str().unwrap();
    let parent = path.parent().unwrap();
    for item in TEST_DIRS {
        // Ensure slashes are there for easier matching
        let test_root = format!("/{}/", item.test_root);
        let mut path_dir = parent.to_str().unwrap().to_string();
        if !path_dir.ends_with('/') {
            path_dir.push('/');
        }
        if let Some(pos) = path_dir.find(&test_root) {
            let parent = PathBuf::from("snapshots")
                .join(item.snapshot_suffix)
                .join(&path_dir[pos + test_root.len()..]);
            return Some((parent, item));
        }
    }
    panic!("Unknown test directory: {path_str}");
}

/// Test parsing all DBC files
#[test]
fn test_if_submodules_are_present() {
    for test in TEST_DIRS {
        let dir = Path::new("./tests/fixtures").join(test.test_root);
        fs::read_dir(&dir).unwrap_or_else(|e| {
            let dir_display = dir.display();
            panic!(
                "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
            )
        });
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file([path]: [&Path; 1]) {
    let Some((
        snapshot_path,
        &TestConfig {
            use_cp1251,
            create_snapshot,
            ..
        },
    )) = get_test_info(path)
    else {
        return;
    };

    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    let buffer = if use_cp1251 {
        decode_cp1252(&buffer)
            .unwrap_or_else(|| panic!("Failed to decode {} as cp1252", path.display()))
    } else {
        Cow::Borrowed(
            std::str::from_utf8(&buffer)
                .unwrap_or_else(|_| panic!("Failed to decode {} as utf-8", path.display())),
        )
    };

    report_test_result(
        Dbc::try_from(buffer.as_ref()),
        snapshot_path,
        file_name,
        create_snapshot,
    );
}

fn report_test_result(
    result: Result<Dbc, Error>,
    snapshot_path: PathBuf,
    file_name: String,
    do_snapshot: bool,
) {
    let do_snapshot = std::env::var("SKIP_INSTA").is_err()
        && (do_snapshot || std::env::var("FORCE_INSTA").is_ok());

    with_settings! {
        {
            omit_expression => true,
            snapshot_path => snapshot_path,
            prepend_module_to_snapshot => false
        },
        {
            match result {
                Ok(result) => {
                    if do_snapshot {
                        assert_yaml_snapshot!(file_name, result);
                    } // else ignore successful results when not snapshotting
                }
                Err(e) => {
                    let file_name = format!("!error___{file_name}");
                    assert_debug_snapshot!(file_name, err_to_str(&e));
                }
            }
        }
    }
}

fn err_to_str<'a>(err: &'a Error) -> &'a str {
    match err {
        Error::Incomplete(_, _) => "incomplete",
        Error::Nom(e) => match e {
            Err::Incomplete(_) => "nom-incomplete",
            Err::Error(_) => "nom-error",
            Err::Failure(_) => "nom-failure",
        },
        Error::MultipleMultiplexors => "multiple multiplexors",
    }
}
