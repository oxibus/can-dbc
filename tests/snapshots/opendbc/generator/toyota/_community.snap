---
source: tests/snapshots.rs
---
version: ""
bit_timing: ~
messages:
  - id:
      Standard: 359
    name: STEERING_IPAS_COMMA
    size: 8
    transmitter:
      NodeName: IPAS
    signals:
      - name: STATE
        multiplexer_indicator: Plain
        start_bit: 7
        size: 4
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
      - name: ANGLE
        multiplexer_indicator: Plain
        start_bit: 3
        size: 12
        byte_order: BigEndian
        value_type: Signed
        factor: 1.5
        offset: 0
        min: -510
        max: 510
        unit: deg
        receivers:
          - XXX
      - name: SET_ME_X10
        multiplexer_indicator: Plain
        start_bit: 23
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: SET_ME_X00
        multiplexer_indicator: Plain
        start_bit: 31
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: DIRECTION_CMD
        multiplexer_indicator: Plain
        start_bit: 38
        size: 2
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
      - name: SET_ME_X40
        multiplexer_indicator: Plain
        start_bit: 47
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: SET_ME_X00_1
        multiplexer_indicator: Plain
        start_bit: 55
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 63
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 512
    name: GAS_COMMAND
    size: 6
    transmitter:
      NodeName: EON
    signals:
      - name: GAS_COMMAND
        multiplexer_indicator: Plain
        start_bit: 7
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 0.159375
        offset: -75.555
        min: 0
        max: 1
        unit: ""
        receivers:
          - INTERCEPTOR
      - name: GAS_COMMAND2
        multiplexer_indicator: Plain
        start_bit: 23
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 0.159375
        offset: -151.111
        min: 0
        max: 1
        unit: ""
        receivers:
          - INTERCEPTOR
      - name: ENABLE
        multiplexer_indicator: Plain
        start_bit: 39
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - INTERCEPTOR
      - name: COUNTER_PEDAL
        multiplexer_indicator: Plain
        start_bit: 35
        size: 4
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - INTERCEPTOR
      - name: CHECKSUM_PEDAL
        multiplexer_indicator: Plain
        start_bit: 47
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - INTERCEPTOR
  - id:
      Standard: 513
    name: GAS_SENSOR
    size: 6
    transmitter:
      NodeName: INTERCEPTOR
    signals:
      - name: INTERCEPTOR_GAS
        multiplexer_indicator: Plain
        start_bit: 7
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - EON
      - name: INTERCEPTOR_GAS2
        multiplexer_indicator: Plain
        start_bit: 23
        size: 16
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - EON
      - name: STATE
        multiplexer_indicator: Plain
        start_bit: 39
        size: 4
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - EON
      - name: COUNTER_PEDAL
        multiplexer_indicator: Plain
        start_bit: 35
        size: 4
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - EON
      - name: CHECKSUM_PEDAL
        multiplexer_indicator: Plain
        start_bit: 47
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - EON
  - id:
      Standard: 35
    name: SECONDARY_STEER_ANGLE
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: ZORRO_STEER
        multiplexer_indicator: Plain
        start_bit: 7
        size: 24
        byte_order: BigEndian
        value_type: Signed
        factor: 0.004901594652
        offset: 0
        min: -500
        max: 500
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 767
    name: SDSU
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: FD_BUTTON
        multiplexer_indicator: Plain
        start_bit: 7
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: STATE
        multiplexer_indicator: Plain
        start_bit: 23
        size: 4
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
comments:
  - Plain:
      comment: "BO_ STEERING_IPAS_COMMA: Copy of msg 614 so we can do angle control while the Park Assist ECU is connected (Panda spoofs 614 with 359 on connector J70). Note that addresses 0x266 and 0x167 are checksum-invariant"
  - Plain:
      comment: "BO_ SECONDARY_STEER_ANGLE: ZSS is a high-precision steering angle sensor that can replace the lower resolution sensor in most TSS1 Toyotas. Learn more: https://github.com/commaai/openpilot/wiki/Toyota-Lexus#zorro-steering-sensor-zss"
  - Plain:
      comment: "BO_ SDSU: The sDSU is a modified DSU for use in TSS-P Toyotas. Learn more: https://github.com/RetroPilot/ocelot/tree/main/firmware/smart_dsu"
  - Signal:
      message_id:
        Standard: 767
      name: FD_BUTTON
      comment: The follow distance button signal as forwarded by the sdsu
value_descriptions:
  - Signal:
      message_id:
        Standard: 513
      name: STATE
      value_descriptions:
        - id: 5
          description: FAULT_TIMEOUT
        - id: 4
          description: FAULT_STARTUP
        - id: 3
          description: FAULT_SCE
        - id: 2
          description: FAULT_SEND
        - id: 1
          description: FAULT_BAD_CHECKSUM
        - id: 0
          description: NO_FAULT
  - Signal:
      message_id:
        Standard: 767
      name: STATE
      value_descriptions:
        - id: 7
          description: STATE_AEB_CTRL
        - id: 6
          description: FAULT_INVALID
        - id: 5
          description: FAULT_TIMEOUT
        - id: 4
          description: FAULT_STARTUP
        - id: 3
          description: FAULT_SCE
        - id: 2
          description: FAULT_SEND
        - id: 1
          description: FAULT_BAD_CHECKSUM
        - id: 0
          description: NO_FAULT
